### PR TITLE
fern: Fourier learnable B + m∈{96,160,256} sweep on best recipe

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The kagent track and willow-pai2-r1 use a **fixed** random Fourier features (RFF) projection: `B ~ N(0, σ²)` is sampled once at init and registered as a non-trainable buffer. This is theoretically optimal under "Fourier features let networks learn high-frequency functions" (Tancik et al. 2020). Two adjacent dimensions remain underexplored on this codebase:

1. **Learnable B** — make the projection matrix `B` a `nn.Parameter` and let SGD adapt it. With nl=3 and only ~100 epochs of total updates, the optimizer might (a) discover task-tuned frequencies and beat fixed RFF, OR (b) collapse to a low-frequency mode and underperform. Both outcomes are diagnostic.
2. **Frequency budget** — `fourier_m=160` was picked on the kagent track via fine sweep at fixed σ=0.7. With nl=3 we have a budget-bound model — more frequencies (m=256) could give the trunk more representational room, but only if the early Fourier embedding isn't already saturating.

This PR co-tests both dimensions: `fourier_features ∈ {fixed, learnable}` × `fourier_m ∈ {96, 160, 256}`, with σ=0.7 held fixed (kagent's known-good).

**Predicted outcome:**
- `learnable, m=160`: most likely winner (≥1 val improvement) IF SGD has enough budget to specialize.
- `learnable, m=256`: tied or slight win if frequencies collapse with smaller m. VRAM likely doubles for the projection.
- `fixed, m=256`: marginal change vs m=160 (most signal already captured).
- `fixed, m=96`: small regression (~+0.5 val) — confirms m floor.

**Predicted delta vs (fixed, m=160) anchor:** −0.5 to −2 val for the best learnable variant; otherwise within noise.

## Instructions

Modify only `target/train.py`. Implement the full bootstrap recipe (same as askeladd's `best-recipe-bootstrap-5seed` PR — see that PR's body for step-by-step), then add the `learnable` Fourier mode and run the sweep.

### Phase 1: Implement bootstrap recipe (see askeladd PR for full spec)

All flags, seed pinning, L1 loss, Fourier PE, SwiGLU, AMP+grad-accum, parametric n_layers/slice_num/n_head.

### Phase 2: Add `learnable` Fourier mode

Replace the `FourierEmbedding` class to support both modes:

```python
class FourierEmbedding(nn.Module):
    def __init__(self, in_dim: int, m: int, sigma: float, learnable: bool = False):
        super().__init__()
        B = torch.randn(in_dim, m) * sigma
        if learnable:
            self.B = nn.Parameter(B)        # SGD-trainable
        else:
            self.register_buffer("B", B)    # frozen RFF
    def forward(self, x):
        proj = x @ self.B
        return torch.cat([torch.sin(proj), torch.cos(proj)], dim=-1)
```

Update the instantiation:

```python
fourier_emb = (
    FourierEmbedding(2, cfg.fourier_m, cfg.fourier_sigma,
                     learnable=(cfg.fourier_features == "learnable")).to(device)
    if cfg.fourier_features in ("fixed", "learnable") else None
)
```

Update the Config flag valid values: `fourier_features: str = "none"` should now accept `"none" | "fixed" | "learnable"`. Document this in a comment near the Config field.

### Phase 3: Pre-sweep verification

ONE `--debug` run at the largest config (m=256, learnable) to confirm VRAM is OK and gradients flow to `B`:

```bash
cd target && CUDA_VISIBLE_DEVICES=0 python train.py --debug \
    --agent willowpai2r4-fern \
    --loss_type l1 --surf_weight 1 --amp true --grad_accum 4 --batch_size 4 \
    --fourier_features learnable --fourier_m 256 --fourier_sigma 0.7 --swiglu true \
    --n_layers 3 --slice_num 8 --n_head 1 --seed 0 \
    --wandb_group fern/fourier-learn-debug --wandb_name fern/debug-verify
```

Verify the W&B Config shows `fourier_features=learnable, fourier_m=256` and that `n_params` rises by ~512 (2*256 = +512 trainable parameters in the Fourier projection).

### Sweep matrix

You have 1 GPU. Run sequentially under `--wandb_group fern/fourier-learnable-m-sweep`.

| Run | `--fourier_features` | `--fourier_m` | seed | `--wandb_name` | Purpose |
|-----|----------------------|---------------|------|----------------|---------|
| 1   | fixed                | 160           | 0    | `fern/fix-m160-s0`     | Anchor (matches askeladd recipe) |
| 2   | fixed                | 160           | 1    | `fern/fix-m160-s1`     | 2-seed anchor |
| 3   | learnable            | 160           | 0    | `fern/learn-m160-s0`   | PRIMARY hypothesis |
| 4   | learnable            | 160           | 1    | `fern/learn-m160-s1`   | 2-seed primary |
| 5   | learnable            | 256           | 0    | `fern/learn-m256-s0`   | More-frequencies probe |
| 6   | fixed                | 256           | 0    | `fern/fix-m256-s0`     | Diagnostic — does m alone help? |
| 7   | fixed                | 96            | 0    | `fern/fix-m96-s0`      | m floor probe |

Priority order: 1, 2, 3, 4 first (the comparison that matters). Then 5, 6, 7 as budget allows.

All runs use:
```bash
cd target && CUDA_VISIBLE_DEVICES=0 python train.py \
    --agent willowpai2r4-fern \
    --loss_type l1 --surf_weight 1 --amp true --grad_accum 4 --batch_size 4 \
    --fourier_sigma 0.7 --swiglu true \
    --n_layers 3 --slice_num 8 --n_head 1 \
    --fourier_features <FF> --fourier_m <M> --seed <S> \
    --wandb_group fern/fourier-learnable-m-sweep \
    --wandb_name fern/<FF>-m<M>-s<S>
```

### Reporting

1. **Per-run table:** ff, m, seed, val_avg, test_avg, best_epoch, epochs, peak VRAM, time/epoch, n_params (Fourier proj contribution), W&B run ID.
2. **2-seed means** for (fixed, 160) anchor and (learnable, 160) primary; std too.
3. **Per-split breakdown** for the best variant (4 val + 4 test, mae_surf_p).
4. **B matrix diagnostics** for learnable runs: distribution of `||B[:, j]||` after training vs at init. Has the optimizer concentrated frequencies? Reduce-mean of `B.abs()` end-vs-start.
5. **Throughput:** does m=256 meaningfully slow training?

### Merge criterion

- If (learnable, m=160) 2-seed mean beats (fixed, m=160) 2-seed mean by ≥ 1.0 val: WIN, merge.
- If within ±1.0: equivalent, prefer fixed for simplicity, but report.
- If learnable regresses by > 1.0: report B-collapse mechanism in writeup, close.

## Baseline

**Track:** Fresh willow-pai2-r4. Default baseline in PR #105 (alphonse).

**Reference (askeladd recipe in PR willowpai2r4-askeladd):** anchor, expected 5-seed mean val ~49–51. Your (fixed, m=160) seeds will reproduce this.

**Reference (kagent PR #39):** at fixed/m=160/σ=0.7, 2-seed mean val 49.443 ± 0.517, test 42.450.

Reproduce the (fixed, m=160) anchor:
```bash
cd target && CUDA_VISIBLE_DEVICES=0 python train.py \
    --agent willowpai2r4-fern --seed 0 \
    --loss_type l1 --surf_weight 1 --amp true --grad_accum 4 --batch_size 4 \
    --fourier_features fixed --fourier_m 160 --fourier_sigma 0.7 --swiglu true \
    --n_layers 3 --slice_num 8 --n_head 1 \
    --wandb_group fern/fourier-learnable-m-sweep --wandb_name fern/fix-m160-s0
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
Closing — created in wrong repo by automation glitch. Will be re-created in morganmcg1/TandemFoilSet-Balanced.
